File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.cpp
class Solution {
public:
    bool pacific[201][201];  
    bool atlantic[201][201];  
    vector<pair<int,int>> dir = {{0,1},{0,-1},{1,0},{-1,0}}; 
    void dfs(int i, int j, int n, int m, char finder, vector<vector<int>>& heights){
        if(i < 0 || j < 0 || i >= n || j >= m) return; 
        
        if(finder == 'P') pacific[i][j] = true;  
        if(finder == 'A') atlantic[i][j] = true; 

        for(pair<int,int>& p : dir){
            int nX = i + p.first; 
            int nY = j + p.second; 

            if(i < 0 || j < 0 || i >= n || j >= m) continue; 
            if(nX < 0 || nY < 0 || nX >= n || nY >= m) continue; 

            if(finder == 'P' && !pacific[nX][nY] && heights[nX][nY] >= heights[i][j]) dfs(nX, nY, n, m, finder, heights); 
            if(finder == 'A' && !atlantic[nX][nY] && heights[nX][nY] >= heights[i][j]) dfs(nX, nY, n, m, finder, heights); 
        }
    }
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        vector<vector<int>> answer; 
        int n = heights.size(); 
        int m = heights[0].size(); 
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(i == 0 || j == 0){
                    dfs(i,j,n,m,'P',heights); 
                }
                if(i == n-1 || j == m-1){
                    dfs(i,j,n,m,'A',heights);
                }
            }
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                //cout << pacific[i][j] << ' '; 
                if(pacific[i][j] && atlantic[i][j]) answer.push_back({i,j}); 
            }
            //cout << endl; 
        }
        return answer; 
    }
};

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    boolean[][] pacific = new boolean[201][201]; 
    boolean[][] atlantic = new boolean[201][201];
    int[][] dir = {{0,1},{0,-1},{1,0},{-1,0}}; 
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> answer = new ArrayList<>(); 
        int m = heights.length, n = heights[0].length; 
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(i == 0 || j == 0){
                    dfs(heights,i,j,'P'); 
                }
                if(i == m-1 || j == n-1){
                    dfs(heights,i,j,'A'); 
                }
            }
        }

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                List<Integer> container = new ArrayList<>(); 

                if(pacific[i][j] && atlantic[i][j]){
                    container.add(i);
                    container.add(j); 
                    answer.add(container); 
                }
            }
        }



        return answer; 
    }

    public void dfs(int[][] heights, int i, int j, char ocean){
        
        if(ocean == 'P') pacific[i][j] = true;
        if(ocean == 'A') atlantic[i][j] = true; 

        for(int[] d : dir){
            int nX = i + d[0]; 
            int nY = j + d[1]; 
            if(nX >= 0 && nY >= 0 && nX < heights.length && nY < heights[0].length && heights[i][j] <= heights[nX][nY]){
                if(ocean == 'P' && pacific[nX][nY] != true ) dfs(heights,nX,nY,ocean); 
                if(ocean == 'A' && atlantic[nX][nY] != true) dfs(heights,nX,nY,ocean); 
            }
        }
    }
}